src/expr/src/scalar/func.rs

```diff
@@ -2499,116 +2499,190 @@ impl BinaryFunc {
+            | SubNumeric
+            | MulInt16
+            | MulInt32
+            | MulInt64
+            | MulUInt16
+            | MulUInt32
+            | MulUInt64
+            | MulFloat32
+            | MulFloat64
+            | MulNumeric
+            | MulInterval
+            | DivInt16
+            | DivInt32
+            | DivInt64
+            | DivUInt16
+            | DivUInt32
+            | DivUInt64
+            | DivFloat32
+            | DivFloat64
+            | DivNumeric
+            | DivInterval
+            | ModInt16
+            | ModInt32
+            | ModInt64
+            | ModUInt16
+            | ModUInt32
+            | ModUInt64
+            | ModFloat32
+            | ModFloat64
+            | ModNumeric
+            | RoundNumeric
+            | Eq
+            | NotEq
+            | Lt
+            | Lte
+            | Gt
+            | Gte
+            | LikeEscape
+            | IsLikeMatch { .. }
+            | ToCharTimestamp
+            | ToCharTimestampTz
+            | DateBinTimestamp
+            | DateBinTimestampTz
+            | ExtractInterval
+            | ExtractTime
+            | ExtractTimestamp
+            | ExtractTimestampTz
+            | ExtractDate
+            | DatePartInterval
+            | DatePartTime
+            | DatePartTimestamp
+            | DatePartTimestampTz
+            | DateTruncTimestamp
+            | DateTruncTimestampTz
+            | DateTruncInterval
+            | TimezoneTimestamp
+            | TimezoneTimestampTz
+            | TimezoneTime { .. }
+            | TimezoneIntervalTimestamp
+            | TimezoneIntervalTimestampTz
+            | TimezoneIntervalTime
+            | TextConcat
+            | JsonbContainsString
+            | JsonbContainsJsonb
+            | MapContainsKey
+            | MapContainsAllKeys
+            | MapContainsAnyKeys
+            | MapContainsMap
+            | ConvertFrom
+            | Left
+            | Position
+            | Right
+            | RepeatString
+            | Trim
+            | TrimLeading
+            | TrimTrailing
+            | EncodedBytesCharLength
+            | ArrayContains
+            | ArrayRemove
+            | ArrayArrayConcat
+            | ListListConcat
+            | ListElementConcat
+            | ElementListConcat
+            | ListRemove
+            | DigestString
+            | DigestBytes
+            | MzRenderTypmod
+            | Encode
+            | Decode
+            | LogNumeric
+            | Power
+            | PowerNumeric
+            | GetByte
+            | RangeContainsElem { .. }
+            | RangeContainsRange { .. }
+            | RangeOverlaps
+            | RangeAfter
+            | RangeBefore
+            | RangeOverleft
+            | RangeOverright
+            | RangeAdjacent
+            | RangeUnion
+            | RangeIntersection
+            | RangeDifference => false,
+
+            IsRegexpMatch { .. }
```

Comment:
You're right---it doesn't.

Actual behavior in `is_like_match_dynamic` and `is_regexp_match_dynamic` is to error on invalid pattern.

I don't know where to look in the docs beyond https://materialize.com/docs/sql/functions/, which document the existence of the `~` opeartor but not its behavior. `LIKE` is similarly not documented in a detailed way (there, at least).

In Postgres, invalid like patterns (`select 'abc' like '\';`) error out; regexes behave similarly (`select 'abc' similar to '*';`).

Should I update `BinaryFunc::output_type` in a fix?

Reply:
Ok, then I'd say
- The behavior is fine, since it matches Postgres (and makes sense, which is also nice :)).
- `IsRegexpMatch` and `IsLikeMatch` should both be `false`.
- The docs are fine. I'd say it's not necessary to document that we error out on an invalid input. I was just worried about the docs possibly incorrectly saying that we output null. (Btw. if the behavior would be outputting null, that would have needed documentation IMO.)
- Yes, please fix `BinaryFunc::output_type`. This can go in this same PR.

Reply:
Haha, always nice to have things make sense! 😁